<a href="https://colab.research.google.com/github/KevinBezerra/Resume/blob/main/Consultoria_Matchbox.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importar bibliotecas

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!pip install unidecode

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 235.9/235.9 KB 8.2 MB/s eta 0:00:00


In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

import json
import unidecode

# Carregar base de dados

In [ ]:
#@markdown Local do arquivo
#file_path = "/content/Base - Dados Clientes - P\xE1gina1.csv" #@param ["/content/drive/MyDrive/arquivos análise /Base - Dados Clientes - Página1.csv", "/content/Base - Dados Clientes - Página1.csv"] {allow-input: true}

In [6]:
 # df = pd.read_csv(file_path)

 df = pd.read_csv("/content/drive/MyDrive/Datasets/Base - Dados Clientes - Página1.csv")

# Visualização básica dos dados

In [7]:
# Exemplos de dados
df.head(5)

Empresa Produto Data de Entrada Data de Saída Tipo Saída Tipo Plano  \
0  4R Serviços  HR Bot      21/03/2021           NaN        NaN      Colab   
1   Alpargatas   Spark      22/10/2022           NaN        NaN     Talent   
2        Alper  HR Bot      21/01/2021           NaN        NaN     Talent   
3       Amaggi  HR Bot      22/08/2022           NaN        NaN     Talent   
4        Ambev   Spark      21/01/2021           NaN        NaN     Talent   

                 CNPJ                                   Razão Social  \
0  24.689.331/0001-49                   CARPLUS SERVICOS AUTOMOTIVOS   
1  61.079.117/0001-05                                ALPARGATAS S.A.   
2  11.721.921/0001-60  ALPER CONSULTORIA E CORRETORA DE SEGUROS S.A.   
3  77.294.254/0001-94            AMAGGI EXPORTACAO E IMPORTACAO LTDA   
4  07.526.557/0001-00                                     AMBEV S.A.   

                                   Natureza Jurídica Data Abertura  \
0  Empresa Individual de Responsabilidade Limitad...    20/04/2016   
1                    2046 | SOCIEDADE ANÔNIMA ABERTA    09/04/1907   
2                           SOCIEDADE ANÔNIMA ABERTA    15/03/2010   
3                      SOCIEDADE EMPRESÁRIA LIMITADA    03/01/1977   
4                    2046 | SOCIEDADE ANÔNIMA ABERTA    19/07/2005   

    Cidade/Estado                      Atividade Econômica Principal  \
0  GUARULHOS - SP  SERVIÇOS DE LANTERNAGEM OU FUNILARIA E PINTURA...   
1  SAO PAULO | SP           PREPARAÇÃO E FIAÇÃO DE FIBRAS DE ALGODÃO   
2  SAO PAULO | SP  CORRETORES E AGENTES DE SEGUROS, DE PLANOS DE ...   
3     CUIABA | MT                        COMÉRCIO ATACADISTA DE SOJA   
4  SAO PAULO | SP                    FABRICAÇÃO DE CERVEJAS E CHOPES   

                             Setor         Tamanho da empresa  \
0        Administração de serviços        51-200 funcionários   
1                       Manufatura   + de 10.001 funcionários   
2                          Seguros       201-500 funcionários   
3                      Agricultura  5.001-10.000 funcionários   
4  Serviços de alimentos e bebidas   + de 10.001 funcionários   

                                            LinkedIn  \
0  https://www.linkedin.com/company/street-servi%...   
1       https://www.linkedin.com/company/alpargatas/   
2  https://www.linkedin.com/company/alper-seguros...   
3           https://www.linkedin.com/company/amaggi/   
4            https://www.linkedin.com/company/ambev/   

                                 Instagram  \
0    https://www.instagram.com/4rservicos/   
1    https://www.instagram.com/alpargatas/   
2  https://www.instagram.com/alperseguros/   
3      https://www.instagram.com/amaggibr/   
4         https://www.instagram.com/ambev/   

                                   Informe Cadastral Unnamed: 17  
0  https://www.informecadastral.com.br/cnpj/4r-se...         NaN  
1  https://www.informecadastral.com.br/cnpj/alpar...         NaN  
2  https://www.informecadastral.com.br/cnpj/alper...         NaN  
3  https://www.informecadastral.com.br/cnpj/amagg...         NaN  
4  https://www.informecadastral.com.br/cnpj/ambev...         NaN

In [8]:
# Tipos dos dados
df.dtypes

Empresa                          object
Produto                          object
Data de Entrada                  object
Data de Saída                    object
Tipo Saída                       object
Tipo Plano                       object
CNPJ                             object
Razão Social                     object
Natureza Jurídica                object
Data Abertura                    object
Cidade/Estado                    object
Atividade Econômica Principal    object
Setor                            object
Tamanho da empresa               object
LinkedIn                         object
Instagram                        object
Informe Cadastral                object
Unnamed: 17                      object
dtype: object

# Entendimento e tratamento detalhado das colunas

Observações:
- Colunas de datas estão como texto (_object_)
- As colunas 'Empresa' e 'Produto' podem ser codificadas para preservar a privacidade dos dados. Pode-ser utilizar, ao invés de 'Empresa', a coluna de 'CNPJ' aplicando a técnica de Shuffling para mascarar o dado real.
- É possível manter apenas uma das 3 colunas, entre 'Empresa', 'Razão Social' e 'CNPJ'
- Cidade e estado podem ser separados
- A coluna 'Tamanho da empresa' pode ser tratada para manter apenas as faixas de quantidades ou utilizar a técnica de Ordinal Encoding
- As colunas 'Linkedin', 'Instagram' e 'Informe Cadastral' podem ser utilizadas para prospectar dados relacionados à quantidade de seguidores, atividade social, situação cadastral, endereço (maior precisão do que cidade), patrimônio, que podem complementar a análise, dependendo do objetivo. No entanto, devem ser utilizadas com cuidado para garantir a privacidade dos dados

In [ ]:
# Verificar as estatísticas básicas das colunas
df.describe(include='all', datetime_is_numeric=True).T

count unique  \
Empresa                          73     73   
Produto                          73      2   
Data de Entrada                  73     23   
Data de Saída                    11      8   
Tipo Saída                       11      2   
Tipo Plano                       73      2   
CNPJ                             72     71   
Razão Social                     72     72   
Natureza Jurídica                72     12   
Data Abertura                    69     68   
Cidade/Estado                    72     27   
Atividade Econômica Principal    68     52   
Setor                            70     33   
Tamanho da empresa               70      7   
LinkedIn                         70     70   
Instagram                        67     67   
Informe Cadastral                68     68   
Unnamed: 17                       2      2   

                                                                             top  \
Empresa                                                              4R Serviços   
Produto                                                                    Spark   
Data de Entrada                                                       21/03/2021   
Data de Saída                                                         22/05/2022   
Tipo Saída                                                            Voluntário   
Tipo Plano                                                                Talent   
CNPJ                                                          60.561.800/0001-03   
Razão Social                                        CARPLUS SERVICOS AUTOMOTIVOS   
Natureza Jurídica                                  SOCIEDADE EMPRESÁRIA LIMITADA   
Data Abertura                                                         08/07/1966   
Cidade/Estado                                                     SAO PAULO | SP   
Atividade Econômica Principal           BANCOS MÚLTIPLOS, COM CARTEIRA COMERCIAL   
Setor                                          Tecnologia, Informação e Internet   
Tamanho da empresa                                      + de 10.001 funcionários   
LinkedIn                       https://www.linkedin.com/company/street-servi%...   
Instagram                                  https://www.instagram.com/4rservicos/   
Informe Cadastral              https://www.informecadastral.com.br/cnpj/4r-se...   
Unnamed: 17                    fATURAMENTO VEM EM NOME DA DIALOG - UM CANAL D...   

                              freq  
Empresa                          1  
Produto                         59  
Data de Entrada                  7  
Data de Saída                    2  
Tipo Saída                       7  
Tipo Plano                      56  
CNPJ                             2  
Razão Social                     1  
Natureza Jurídica               20  
Data Abertura                    2  
Cidade/Estado                   31  
Atividade Econômica Principal    4  
Setor                            5  
Tamanho da empresa              28  
LinkedIn                         1  
Instagram                        1  
Informe Cadastral                1  
Unnamed: 17                      1

count do CNPJ deu	72 e unique 71

In [ ]:
# Contar duplicadas de CNPJ para verificar possibilidade de cadastro duplicado. Usar CNPJ pois existe o risco de 2 empresas terem um nome muito parecido.
df.groupby('CNPJ').size().sort_values(ascending=False)

CNPJ
60.561.800/0001-03     2
 61.064.838/0001-33    1
58.069.360/0001-20     1
51.855.716/0001-01     1
50.746.577/0001-15     1
                      ..
11.725.176/0001-27     1
11.721.921/0001-60     1
11.525.226/0001-22     1
10.588.595/0010-92     1
92.754.738/0001-62     1
Length: 71, dtype: int64

 Identifiquei 2 empresas com mesmo cnpj 
- 60.561.800/0001-03

In [ ]:
# já identifiquei 2 empresas com mesmo cnpj : '60.561.800/0001-03'
df.loc[df['CNPJ']=='60.561.800/0001-03', ['Empresa', 'Produto', 'CNPJ', 'Razão Social', 'Natureza Jurídica', 'Informe Cadastral']]

Empresa Produto                CNPJ             Razão Social  \
51     Novelis   Spark  60.561.800/0001-03  NOVELIS DO BRASIL LTDA.   
52  Odontoprev  HR Bot  60.561.800/0001-03          ODONTOPREV S.A.   

                  Natureza Jurídica  \
51    SOCIEDADE EMPRESÁRIA LIMITADA   
52  2046 | SOCIEDADE ANÔNIMA ABERTA   

                                    Informe Cadastral  
51  https://www.situacaocadastral.info/cnpj/noveli...  
52  https://www.informecadastral.com.br/cnpj/odont...

In [ ]:
# Checar o link do 'Informe Cadastral' para verificar qual está incorreto
print(df.iloc[51]['Informe Cadastral'])
print(df.iloc[52]['Informe Cadastral'])

https://www.situacaocadastral.info/cnpj/novelis-do-brasil-ltda-60561800000103
https://www.informecadastral.com.br/cnpj/odontoprev-sa-58119199000151



- São empresas diferentes. É possível verificar que a Odontoprev foi cadastrada com CNPJ incorreto.
- Vou dar input correto no cnpj, para poder prosseguir com análise.
- Pode se excluir as colunas 'Empresa' e 'Razão Social' após codificar o CNPJ

In [ ]:
# Vou dar input correto no cnpj, para poder prosseguir com análise.
df.loc[52, "CNPJ"] = "58.119.199/0001-51"
df.loc[df['CNPJ']=='58.119.199/0001-51', ['Empresa', 'Produto', 'CNPJ', 'Razão Social', 'Natureza Jurídica', 'Informe Cadastral']]

Empresa Produto                CNPJ     Razão Social  \
52  Odontoprev  HR Bot  58.119.199/0001-51  ODONTOPREV S.A.   

                  Natureza Jurídica  \
52  2046 | SOCIEDADE ANÔNIMA ABERTA   

                                    Informe Cadastral  
52  https://www.informecadastral.com.br/cnpj/odont...

In [ ]:
# Aplicar label encoding no CNPJ usando sklearn LabelEncoder para preservar a privacidade dos dados
le = LabelEncoder()
le.fit(df['CNPJ'])

 # posteriormente é possível resgatar o CNPJ original utilizando: df['CNPJ'] = le.inverse_transform(df['CNPJ cod'])
df['CNPJ cod'] = le.transform(df['CNPJ'])

In [ ]:
# Remover as colunas 'Empresa', 'Razão Social' e 'CNPJ' para manter apenas 'CNPJ cod'
df.drop(columns=['Empresa', 'Razão Social', 'CNPJ'], inplace=True)

In [ ]:
df_Copia = df # tirando ima copia
df.head(5)

Produto Data de Entrada Data de Saída Tipo Saída Tipo Plano  \
0  HR Bot      21/03/2021           NaN        NaN      Colab   
1   Spark      22/10/2022           NaN        NaN     Talent   
2  HR Bot      21/01/2021           NaN        NaN     Talent   
3  HR Bot      22/08/2022           NaN        NaN     Talent   
4   Spark      21/01/2021           NaN        NaN     Talent   

                                   Natureza Jurídica Data Abertura  \
0  Empresa Individual de Responsabilidade Limitad...    20/04/2016   
1                    2046 | SOCIEDADE ANÔNIMA ABERTA    09/04/1907   
2                           SOCIEDADE ANÔNIMA ABERTA    15/03/2010   
3                      SOCIEDADE EMPRESÁRIA LIMITADA    03/01/1977   
4                    2046 | SOCIEDADE ANÔNIMA ABERTA    19/07/2005   

    Cidade/Estado                      Atividade Econômica Principal  \
0  GUARULHOS - SP  SERVIÇOS DE LANTERNAGEM OU FUNILARIA E PINTURA...   
1  SAO PAULO | SP           PREPARAÇÃO E FIAÇÃO DE FIBRAS DE ALGODÃO   
2  SAO PAULO | SP  CORRETORES E AGENTES DE SEGUROS, DE PLANOS DE ...   
3     CUIABA | MT                        COMÉRCIO ATACADISTA DE SOJA   
4  SAO PAULO | SP                    FABRICAÇÃO DE CERVEJAS E CHOPES   

                             Setor         Tamanho da empresa  \
0        Administração de serviços        51-200 funcionários   
1                       Manufatura   + de 10.001 funcionários   
2                          Seguros       201-500 funcionários   
3                      Agricultura  5.001-10.000 funcionários   
4  Serviços de alimentos e bebidas   + de 10.001 funcionários   

                                            LinkedIn  \
0  https://www.linkedin.com/company/street-servi%...   
1       https://www.linkedin.com/company/alpargatas/   
2  https://www.linkedin.com/company/alper-seguros...   
3           https://www.linkedin.com/company/amaggi/   
4            https://www.linkedin.com/company/ambev/   

                                 Instagram  \
0    https://www.instagram.com/4rservicos/   
1    https://www.instagram.com/alpargatas/   
2  https://www.instagram.com/alperseguros/   
3      https://www.instagram.com/amaggibr/   
4         https://www.instagram.com/ambev/   

                                   Informe Cadastral Unnamed: 17  CNPJ cod  
0  https://www.informecadastral.com.br/cnpj/4r-se...         NaN        36  
1  https://www.informecadastral.com.br/cnpj/alpar...         NaN        60  
2  https://www.informecadastral.com.br/cnpj/alper...         NaN        22  
3  https://www.informecadastral.com.br/cnpj/amagg...         NaN        65  
4  https://www.informecadastral.com.br/cnpj/ambev...         NaN        11

In [ ]:
# excluído colunas que não será utilizado
df = df.drop(columns=['LinkedIn', 'Instagram','Informe Cadastral'])

# tranformado em formato data 
df['Data de Entrada'] = pd.to_datetime(df['Data de Entrada'], format='%d/%m/%Y')
df['Data de Saída'] = pd.to_datetime(df['Data de Saída'], format='%d/%m/%Y')
df['Data Abertura'] = pd.to_datetime(df['Data Abertura'], format='%d/%m/%Y')

In [ ]:
# Confirmar tipo de dados alterado
df[['Data de Entrada', 'Data de Saída', 'Data Abertura']].dtypes

Data de Entrada    datetime64[ns]
Data de Saída      datetime64[ns]
Data Abertura      datetime64[ns]
dtype: object

In [ ]:
# Verificar as estatísticas das colunas de data
df[['Data de Entrada', 'Data de Saída', 'Data Abertura']].describe(datetime_is_numeric=True)

Data de Entrada                  Data de Saída  \
count                             73                             11   
mean   2022-01-21 00:39:27.123287552  2022-06-13 19:38:10.909090816   
min              2021-01-21 00:00:00            2022-01-22 00:00:00   
25%              2021-07-21 00:00:00            2022-04-06 12:00:00   
50%              2022-03-22 00:00:00            2022-05-22 00:00:00   
75%              2022-07-22 00:00:00            2022-09-06 12:00:00   
max              2022-11-22 00:00:00            2022-10-22 00:00:00   

             Data Abertura  
count                   69  
mean   1994-11-04 08:00:00  
min    1907-04-09 00:00:00  
25%    1977-09-19 00:00:00  
50%    2001-06-13 00:00:00  
75%    2010-09-22 00:00:00  
max    2020-10-13 00:00:00

Observações:
- A coluna 'Data de Saída' possui poucos dados preenchidos, o que pode indicar que a maioria das empresas ainda estão ativas (não saíram)
- A coluna 'Data Abertura' possui dados faltando, o que pode ser completado através de busca no 'Informe Cadastral'

In [ ]:
# Calcular a idade da empresa e o tempo de permanência ativa
today = np.datetime64('today', 'D')
df['Idade da Empresa (anos)'] = (today - df['Data Abertura']) / np.timedelta64(1, 'Y')
df['Idade da Empresa (anos)'] = df['Idade da Empresa (anos)'].apply(np.round).astype('Int64', errors='ignore')
df['Tempo Permanência Ativa (meses)'] = (df['Data de Saída'].fillna(today) - df['Data de Entrada']) / np.timedelta64(1, 'M') # Preenchendo data de saída em branco com 'today' para calcular o tempo até hoje as empresas que não saíram
df['Tempo Permanência Ativa (meses)'] = df['Tempo Permanência Ativa (meses)'].apply(np.round).astype('Int64', errors='ignore')

In [ ]:
# Checar colunas criadas
df[['Data Abertura', 'Data de Entrada', 'Data de Saída', 'Idade da Empresa (anos)', 'Tempo Permanência Ativa (meses)']].head(2)

Data Abertura Data de Entrada Data de Saída  Idade da Empresa (anos)  \
0    2016-04-20      2021-03-21           NaT                        7   
1    1907-04-09      2022-10-22           NaT                      116   

   Tempo Permanência Ativa (meses)  
0                               22  
1                                3

In [ ]:
# Verificar as estatísticas das colunas de datas novamente, considerando as que foram criadas

df[['Data de Entrada', 'Data de Saída', 'Data Abertura', 'Idade da Empresa (anos)', 'Tempo Permanência Ativa (meses)']].describe(datetime_is_numeric=True)

Data de Entrada                  Data de Saída  \
count                             73                             11   
mean   2022-01-21 00:39:27.123287552  2022-06-13 19:38:10.909090816   
min              2021-01-21 00:00:00            2022-01-22 00:00:00   
25%              2021-07-21 00:00:00            2022-04-06 12:00:00   
50%              2022-03-22 00:00:00            2022-05-22 00:00:00   
75%              2022-07-22 00:00:00            2022-09-06 12:00:00   
max              2022-11-22 00:00:00            2022-10-22 00:00:00   
std                              NaN                            NaN   

             Data Abertura  Idade da Empresa (anos)  \
count                   69                69.000000   
mean   1994-11-04 08:00:00                28.072464   
min    1907-04-09 00:00:00                 2.000000   
25%    1977-09-19 00:00:00                12.000000   
50%    2001-06-13 00:00:00                22.000000   
75%    2010-09-22 00:00:00                45.000000   
max    2020-10-13 00:00:00               116.000000   
std                    NaN                21.201636   

       Tempo Permanência Ativa (meses)  
count                        73.000000  
mean                         10.863014  
min                           2.000000  
25%                           6.000000  
50%                           9.000000  
75%                          15.000000  
max                          24.000000  
std                           6.250145

Observações:
- Embora todas as empresas tenham mais que 2 anos de idade (28 anos na média), nenhuma permanece mais que 2 anos

In [ ]:
# Inserir uma coluna de status: 'ativo' para as que não saíram e 'inativo' para as que já saíram
df['Status'] = df['Data de Saída'].isna().map(lambda x: 'Ativo' if x else 'Inativo')

In [ ]:
# Verificar se mantem inativo as empresas com 'Data de Saída'

df["Data de Saída"] = pd.to_datetime(df["Data de Saída"], errors="coerce")
df[['Data de Saída', 'Status']].drop_duplicates()

Data de Saída   Status
0            NaT    Ativo
7     2022-05-22  Inativo
12    2022-01-22  Inativo
21    2022-04-22  Inativo
32    2022-10-22  Inativo
33    2022-07-22  Inativo
42    2022-08-22  Inativo
54    2022-09-22  Inativo
65    2022-03-22  Inativo

In [ ]:
# Verificar se todas as empresas com data de saída possuem o tipo de saída registrado
df[['Data de Saída', 'Tipo Saída']].describe()

<ipython-input-122-a9a7f312f32d>:2: FutureWarning:

Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.



Data de Saída  Tipo Saída
count                    11          11
unique                    8           2
top     2022-05-22 00:00:00  Voluntário
freq                      2           7
first   2022-01-22 00:00:00         NaN
last    2022-10-22 00:00:00         NaN

- Todas as empresas com saída possuem tipo de saída registrado

A coluna 'Cidade/Estado' apresenta ter dois dados juntos, porém para poder ter uma análise melhor vou separar em colunas distinta 

In [ ]:
df['Cidade/Estado'].unique()

array(['GUARULHOS - SP', 'SAO PAULO | SP', 'CUIABA | MT',
       'BELO HORIZONTE | MG', 'SAN FRANCISCO | CA', 'BRASILIA | DF',
       'BARUERI | SP', 'CAMPINAS | SP', 'RIO DE JANEIRO | RJ', nan,
       'JABOATAO DOS GUARARAPES | PE', 'LONDRINA | PR', 'GOIANIA | GO',
       'CAXIAS DO SUL | RS', 'QUERENCIA | MT',
       'SAO JOSE DOS PINHAIS | PR', 'BLUMENAU | SC', 'PORTO ALEGRE | RS',
       'MOJI MIRIM | SP', 'OSASCO | SP', 'BARRETOS | SP', 'BELEM | PA',
       'ANDRADINA | SP', 'FLORIANOPOLIS | SC',
       'SAO JOSE DO RIO PRETO | SP', 'UBERLANDIA | MG', 'SUZANO | SP',
       'JAGUARIUNA | SP'], dtype=object)

In [ ]:
# Separar cidade e estado
def split_city_state(value):
  value = str(value).split(' | ')
  if len(value) == 1:
    value = value[0].split(' - ')
  if len(value) == 1:
    return value[0], 'nan'
  return value[0], value[1]

# Separar cidade e estado
df['Cidade'] = df['Cidade/Estado'].apply(lambda x: split_city_state(x)[0])
df['Estado'] = df['Cidade/Estado'].apply(lambda x: split_city_state(x)[1])

# Remover coluna 'Cidade/Estado'
df.drop('Cidade/Estado', axis=1, inplace=True)

# Checar alterações se foram corretas
df.groupby(['Estado', 'Cidade']).size()

Estado  Cidade                 
CA      SAN FRANCISCO               1
DF      BRASILIA                    3
GO      GOIANIA                     2
MG      BELO HORIZONTE              3
        UBERLANDIA                  1
MT      CUIABA                      1
        QUERENCIA                   1
PA      BELEM                       1
PE      JABOATAO DOS GUARARAPES     1
PR      LONDRINA                    1
        SAO JOSE DOS PINHAIS        1
RJ      RIO DE JANEIRO              4
RS      CAXIAS DO SUL               1
        PORTO ALEGRE                2
SC      BLUMENAU                    2
        FLORIANOPOLIS               1
SP      ANDRADINA                   1
        BARRETOS                    1
        BARUERI                     4
        CAMPINAS                    2
        GUARULHOS                   1
        JAGUARIUNA                  1
        MOJI MIRIM                  1
        OSASCO                      2
        SAO JOSE DO RIO PRETO       1
        SAO PAULO 

In [ ]:
# Carregar propriedades das cidades brasileiras
# with open ('/content/Propriedades - Cidades BR.json', 'r') as f:
with open ('/content/drive/MyDrive/arquivos análise /Projeto17/Documento de Matheus', 'r') as f:
  cities_data = json.load(f)

In [ ]:
# Remover acentos e tansformar para minúsculo o nome das cidades
for idx, city_data in enumerate(cities_data):
  cities_data[idx]['Cidade min'] = unidecode.unidecode(city_data['Cidade'].lower())

In [ ]:
# Buscar dados da cidade
def search_city_data(value):
  city_data = [x for x in cities_data if x['Cidade min'] == unidecode.unidecode(value).lower()]
  if len(city_data) > 0:
    return city_data[0]
  return {'ID cidade IBGE':np.nan, 'UF': np.nan, 'População IBGE 2020': np.nan, 'Latitude': np.nan, 'Longitude': np.nan}

df['Latitude'] = df['Cidade'].apply(lambda x: search_city_data(x)['Latitude'])
df['Longitude'] = df['Cidade'].apply(lambda x: search_city_data(x)['Longitude'])
df['População'] = df['Cidade'].apply(lambda x: search_city_data(x)['População IBGE 2020'])

In [ ]:
# Checar colunas inseridas
df.groupby(['Estado', 'Cidade', 'Latitude', 'Longitude', 'População']).size()

Estado  Cidade                   Latitude    Longitude   População 
DF      BRASILIA                 -15.794087  -47.887905  3055149.0      3
GO      GOIANIA                  -16.673310  -49.255814  1536097.0      2
MG      BELO HORIZONTE           -19.937524  -43.926453  2521564.0      3
        UBERLANDIA               -18.918999  -48.277950  699097.0       1
MT      CUIABA                   -15.569989  -56.073252  618124.0       1
        QUERENCIA                -12.592300  -52.209892  17937.0        1
PA      BELEM                    -1.459845   -48.487826  1499641.0      1
PE      JABOATAO DOS GUARARAPES  -8.165391   -34.921173  706867.0       1
PR      LONDRINA                 -23.312156  -51.162773  575377.0       1
        SAO JOSE DOS PINHAIS     -25.537702  -49.200082  329058.0       1
RJ      RIO DE JANEIRO           -22.876652  -43.227875  6747815.0      4
RS      CAXIAS DO SUL            -29.166899  -51.178860  517451.0       1
        PORTO ALEGRE             -30.030037 

In [ ]:
# transformando todos os NaN em 0
df = df.replace(np.nan,0)

In [ ]:
df.head(5)

Produto Data de Entrada Data de Saída Tipo Saída Tipo Plano  \
0  HR Bot      2021-03-21             0          0      Colab   
1   Spark      2022-10-22             0          0     Talent   
2  HR Bot      2021-01-21             0          0     Talent   
3  HR Bot      2022-08-22             0          0     Talent   
4   Spark      2021-01-21             0          0     Talent   

                                   Natureza Jurídica        Data Abertura  \
0  Empresa Individual de Responsabilidade Limitad...  2016-04-20 00:00:00   
1                    2046 | SOCIEDADE ANÔNIMA ABERTA  1907-04-09 00:00:00   
2                           SOCIEDADE ANÔNIMA ABERTA  2010-03-15 00:00:00   
3                      SOCIEDADE EMPRESÁRIA LIMITADA  1977-01-03 00:00:00   
4                    2046 | SOCIEDADE ANÔNIMA ABERTA  2005-07-19 00:00:00   

                       Atividade Econômica Principal  \
0  SERVIÇOS DE LANTERNAGEM OU FUNILARIA E PINTURA...   
1           PREPARAÇÃO E FIAÇÃO DE FIBRAS DE ALGODÃO   
2  CORRETORES E AGENTES DE SEGUROS, DE PLANOS DE ...   
3                        COMÉRCIO ATACADISTA DE SOJA   
4                    FABRICAÇÃO DE CERVEJAS E CHOPES   

                             Setor         Tamanho da empresa Unnamed: 17  \
0        Administração de serviços        51-200 funcionários           0   
1                       Manufatura   + de 10.001 funcionários           0   
2                          Seguros       201-500 funcionários           0   
3                      Agricultura  5.001-10.000 funcionários           0   
4  Serviços de alimentos e bebidas   + de 10.001 funcionários           0   

   CNPJ cod  Idade da Empresa (anos)  Tempo Permanência Ativa (meses) Status  \
0        36                        7                               22  Ativo   
1        60                      116                                3  Ativo   
2        22                       13                               24  Ativo   
3        65                       46                                5  Ativo   
4        11                       17                               24  Ativo   

      Cidade Estado   Latitude  Longitude   População  
0  GUARULHOS     SP -23.468506 -46.531084   1392121.0  
1  SAO PAULO     SP -23.567387 -46.570383  12325232.0  
2  SAO PAULO     SP -23.567387 -46.570383  12325232.0  
3     CUIABA     MT -15.569989 -56.073252    618124.0  
4  SAO PAULO     SP -23.567387 -46.570383  12325232.0

In [ ]:
df['Tamanho da empresa'].unique()

array(['51-200 funcionários', '+ de 10.001 funcionários',
       '201-500 funcionários', '5.001-10.000 funcionários',
       '1.001-5.000 funcionários', '501-1.000 funcionários', 0,
       '2-10 funcionários'], dtype=object)

 É possível extrair a quantidade de funcionários de texto para intervalos, para possibilitar a classificação dos dados

In [ ]:
# Transformar o valor de texto em intervalos
def process_company_size(value):
  value = str(value).replace('funcionários', '')
  if '-' in value:
    l, r = value.split('-')
    l, r = l.replace('.', ''), r.replace('.', '')
    return pd.Interval(left=int(l), right=int(r))
  if ' de ' in value:
    r, l = value.split(' de ')
    l, r = l.replace('.', ''), r.replace('.', '')
    return pd.Interval(left=int(l), right=np.inf)
  return pd.Interval(left=0, right=0)

# Transformar intervalo no valor médio de funcionários
def company_size_to_sequence(value, base_list):
  return int((value.left + (value.right if value.right != np.inf else value.left)) / 2)

# Criar uma lista classificada dos intervalos
intervals_list = sorted(map(process_company_size, df['Tamanho da empresa'].astype("str").unique()))

# Processar a coluna tamanho da empresa
df['Tamanho da empresa old'] = df['Tamanho da empresa']
df['Tamanho da empresa'] = df['Tamanho da empresa'].apply(process_company_size)
df['Tamanho da empresa cod'] = df['Tamanho da empresa'].apply(lambda x: company_size_to_sequence(x, intervals_list))

df['Tamanho da empresa'] = df['Tamanho da empresa'].astype(str)


df['Tamanho da empresa'] = df['Tamanho da empresa'].replace({'(10001.0, inf]': '+10001',
                                                              '(1001.0, 5000.0]': '1000 a 5000',
                                                              '(201.0, 500.0]': '200 a 500',
                                                              '(5001.0, 10000.0]': '5000 a 10000',
                                                              '(51.0, 200.0]': '51 a 200',
                                                              '(501.0, 1000.0]': '500 a 1000',
                                                              '(0.0, 0.0]': 'Sem informações',
                                                              '(2.0, 10.0]':  '2 a 10'})

In [ ]:
# Checar quantidade por natureza jurídica
df.groupby('Natureza Jurídica').size().sort_values()

Natureza Jurídica
0                                                                            1
ASSOCIAÇÃO PRIVADA                                                           1
COMÉRCIO ATACADISTA DE ENERGIA ELÉTRICA                                      1
EMPRESA DOMICILIADA NO EXTERIOR                                              1
Empresa Individual de Responsabilidade Limitada (de Natureza Empresária)     1
SOCIEDADE EMPRESARIA LIMITADA                                                1
EMPRESA INDIVIDUAL DE RESPONSABILIDADE LIMITADA (DE NATUREZA EMPRESÁRIA)     2
2054 | SOCIEDADE ANÔNIMA FECHADA                                             5
SOCIEDADE ANÔNIMA FECHADA                                                    6
2062 | SOCIEDADE EMPRESÁRIA LIMITADA                                         7
2046 | SOCIEDADE ANÔNIMA ABERTA                                             11
SOCIEDADE ANÔNIMA ABERTA                                                    16
SOCIEDADE EMPRESÁRIA LIMITADA     

- A coluna possui dados semelhantes com escritas diferentes que podem ser agrupadas


In [ ]:
# Criar dicionário de substituição
legal_nature_replace_dict = {
np.nan: 'nan',
'Empresa Individual de Responsabilidade Limitada (de Natureza Empresária)': 'EIRL',
'2046 | SOCIEDADE ANÔNIMA ABERTA': 'SAA',
'SOCIEDADE ANÔNIMA ABERTA': 'SAA',
'SOCIEDADE EMPRESÁRIA LIMITADA': 'SEL',
'SOCIEDADE ANÔNIMA FECHADA': 'SAF',
'EMPRESA DOMICILIADA NO EXTERIOR': 'EDExt',
'EMPRESA INDIVIDUAL DE RESPONSABILIDADE LIMITADA (DE NATUREZA EMPRESÁRIA)': 'EIRL',
'SOCIEDADE EMPRESARIA LIMITADA': 'SEL',
'ASSOCIAÇÃO PRIVADA': 'AP',
'COMÉRCIO ATACADISTA DE ENERGIA ELÉTRICA': 'CAEE',
'2062 | SOCIEDADE EMPRESÁRIA LIMITADA': 'SEL',
'2054 | SOCIEDADE ANÔNIMA FECHADA': 'SAF'
}

# Substituir valores
df['Natureza Jurídica'].replace(legal_nature_replace_dict, inplace=True)
df.groupby('Natureza Jurídica').size().sort_values()

Natureza Jurídica
0         1
AP        1
CAEE      1
EDExt     1
EIRL      3
SAF      11
SAA      27
SEL      28
dtype: int64

# importar bibliotecas para gráficos 

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.io as pio
pio.templates
from plotly.subplots import make_subplots
import plotly.express as px
import plotly.graph_objects as go

#Análise e gráficos

In [ ]:
df_Produto = df.groupby('Produto').size().sort_values(ascending=False).to_frame()
total = df_Produto.sum()[0]
df_Produto['percentage'] = df_Produto.apply(lambda x: (x/total) * 100, axis=1)
df_Produto['percentage'] = df_Produto['percentage'].apply(lambda x: '{:.2f}%'.format(x))
df_Produto = df_Produto.reset_index()
df_Produto.columns = ["Produtos", "Quantidade", "percentage"]
df_Produto

Produtos  Quantidade percentage
0    Spark          59     80.82%
1   HR Bot          14     19.18%

Produto mais utilizado é Spark com 80,8% contra HR Bot 19,2%

In [ ]:
fig = px.pie(df, names='Produto', labels='Produto', title='Proporção por tipo de produto', width=600)
fig.update_layout(paper_bgcolor='#333333')
fig.update_layout(title_font_color='#FFFFFF')
fig.update_layout(legend_font_color='#FFFFFF')
fig.update_layout(
    title=dict(font=dict(size=20)),
    legend=dict(font=dict(size=20))
)
fig.update_traces(textfont=dict(size=20))
fig.show()

In [ ]:
df_Planos  = df.groupby('Tipo Plano').size().sort_values(ascending=False).to_frame()
total = df_Planos.sum()[0]
df_Planos['percentage'] = df_Planos.apply(lambda x: (x/total) * 100, axis=1)
df_Planos['percentage'] = df_Planos['percentage'].apply(lambda x: '{:.2f}%'.format(x))
df_Planos = df_Planos.reset_index()
df_Planos.columns = ["Tipo Plano", "Quantidade", "percentage"]
df_Planos

Tipo Plano  Quantidade percentage
0     Talent          56     76.71%
1      Colab          17     23.29%

Plano mais utilizado é Talent com 76,7% contra Colab: 23,3%

In [ ]:
fig = px.pie(df, names='Tipo Plano', labels='Tipo Plano', title='Proporção por Tipo Plano', width=600)
fig.update_layout(paper_bgcolor='#333333')
fig.update_layout(title_font_color='#FFFFFF')
fig.update_layout(legend_font_color='#FFFFFF')
fig.update_layout(
    title=dict(font=dict(size=20)),
    legend=dict(font=dict(size=20))
)
fig.update_traces(textfont=dict(size=20))
fig.show()

In [ ]:
df_Pro_pla = df.groupby(['Produto', 'Tipo Plano']).size().sort_values(ascending=False).to_frame()
total = df_Pro_pla.sum()[0]
df_Pro_pla['percentage'] = df_Pro_pla.apply(lambda x: (x/total) * 100, axis=1)
df_Pro_pla['percentage'] = df_Pro_pla['percentage'].apply(lambda x: '{:.2f}%'.format(x))
df_Pro_pla = df_Pro_pla.reset_index()
df_Pro_pla.columns = ["Produto", "Tipo Plano", "Quantidade", "percentage"]
df_Pro_pla

Produto Tipo Plano  Quantidade percentage
0   Spark     Talent          46     63.01%
1   Spark      Colab          13     17.81%
2  HR Bot     Talent          10     13.70%
3  HR Bot      Colab           4      5.48%

Nesse gráfico podemos ver melhor distribuição dos planos por produtos

Spark 
*   Talent 62,87% 
*   Colab  17,81%

HR Bot 
*  Talent  13,70% 
*  Colab 5,48%


In [ ]:
# Cria o gráfico de histograma
fig = px.histogram(df, x='Produto', color='Tipo Plano' ,   title='Proporção de produtos por tipo de Plano', width=600, template='plotly_dark')
fig.update_layout(legend_font_color='#FFFFFF')
fig.update_layout(
    title=dict(font=dict(size=15)),
    legend=dict(font=dict(size=18))
)
fig.update_traces(textfont=dict(size=15))
fig.show()

Com análise anterior sabermos que plano mais utilizado é talent ambos produtos, porém como podemos analisar no gráfico histograma, HR Bot Talemt não conseguir alcançar colab do Spark 

In [ ]:
df_Estado = df.groupby('Estado').size().sort_values(ascending=False).to_frame()
total = df_Estado.sum()[0]
df_Estado['percentage'] = df_Estado.apply(lambda x: (x/total) * 100, axis=1)
df_Estado['percentage'] = df_Estado['percentage'].apply(lambda x: '{:.2f}%'.format(x))
df_Estado = df_Estado.reset_index()
df_Estado.columns = ["Estado","Quantidade", "percentage"]
df_Estado

Estado  Quantidade percentage
0      SP          46     63.01%
1      MG           4      5.48%
2      RJ           4      5.48%
3      DF           3      4.11%
4      RS           3      4.11%
5      SC           3      4.11%
6      GO           2      2.74%
7      MT           2      2.74%
8      PR           2      2.74%
9      CA           1      1.37%
10     PA           1      1.37%
11     PE           1      1.37%
12    nan           1      1.37%

Podemos perceber que tem vários estados do Brasil, porém concentração da base se concentrar em SP 63,0% e restante entre 5,5% a 1,4%.

*  Um ponto de atenção, infelizmente tem 1 cliente com os dados faltante

In [ ]:
fig = px.bar(df_Estado, x='Estado', y='Quantidade')
fig.update_layout(template='plotly_dark')
# Calcula a porcentagem de cada barra
y = df_Estado['Quantidade']
y_total = y.sum()
y_percent = y / y_total * 100
fig.update_traces(
    text=np.round(y_percent,1).astype(str) + '%', # formata a porcentagem com duas casas decimais e o símbolo "%"
    texttemplate='%{text:.s}', # formata o texto com duas casas decimais
    textposition='outside' # posiciona o texto da barra
)
fig.show()

In [ ]:
df_Cidade = df.groupby('Cidade').size().sort_values(ascending=False).to_frame()
total = df_Cidade.sum()[0]
df_Cidade['percentage'] = df_Cidade.apply(lambda x: (x/total) * 100, axis=1)
df_Cidade['percentage'] = df_Cidade['percentage'].apply(lambda x: '{:.2f}%'.format(x))
df_Cidade = df_Cidade.reset_index()
df_Cidade.columns = ["Cidade","Quantidade", "percentage"]
df_Cidade.head(11)

Cidade  Quantidade percentage
0        SAO PAULO          31     42.47%
1          BARUERI           4      5.48%
2   RIO DE JANEIRO           4      5.48%
3   BELO HORIZONTE           3      4.11%
4         BRASILIA           3      4.11%
5     PORTO ALEGRE           2      2.74%
6         BLUMENAU           2      2.74%
7         CAMPINAS           2      2.74%
8           OSASCO           2      2.74%
9          GOIANIA           2      2.74%
10   SAN FRANCISCO           1      1.37%

In [ ]:
fig = px.bar(df_Cidade, x='Cidade', y='Quantidade')
fig.update_layout(template='plotly_dark')
# Calcula a porcentagem de cada barra
y = df_Cidade['Quantidade']
y_total = y.sum()
y_percent = y / y_total * 100
fig.update_traces(
    text=np.round(y_percent,1).astype(str) + '%', # formata a porcentagem com duas casas decimais e o símbolo "%"
    texttemplate='%{text:.s}', # formata o texto com duas casas decimais
    textposition='outside' # posiciona o texto da barra - 'inside' -  dentro da barra
)
fig.show()

Como analisarmos anteriormente SP apresentou ter maior representação da base e Confirmamos agora com Cidade de São Paulo 42,47%

In [ ]:
fig = px.density_mapbox(df, lat='Latitude', lon='Longitude', radius=10,
                        center=dict(lat=-17.14, lon=-57.33), zoom=2.9,
                        mapbox_style="carto-positron")

fig.update_layout(margin={"r":0,"t":0,"l":0,"b":0})
fig.show()

In [ ]:
df_Setor = df.groupby('Setor').size().sort_values(ascending=False).to_frame()
total = df_Setor.sum()[0]
df_Setor['percentage'] = df_Setor.apply(lambda x: (x/total) * 100, axis=1)
df_Setor['percentage'] = df_Setor['percentage'].apply(lambda x: '{:.2f}%'.format(x))
df_Setor1 = df_Setor.reset_index()
df_Setor1.columns = ["Setor","Quantidade", "percentage"]
df_Setor1.head(10)

Setor  Quantidade percentage
0          Tecnologia, Informação e Internet           5      6.85%
1               Serviços e consultoria de TI           5      6.85%
2                                 Manufatura           5      6.85%
3                                  Mineração           4      5.48%
4                                     Varejo           4      5.48%
5                             Petróleo e gás           4      5.48%
6                Desenvolvimento de software           4      5.48%
7  Fabricação de produtos de higiene pessoal           3      4.11%
8                       Serviços financeiros           3      4.11%
9                  Administração de serviços           3      4.11%

Setores de atuação das empresas estão bem distribuídos.

*   Um ponto de atenção, se junta setor 1 ,2 e 3 ambos do seguimento de tecnológia, terá um setor que destacará com 19,18


In [ ]:
fig = px.histogram(df_Setor1, x='Setor' , y= 'Quantidade')
fig.update_layout(template='plotly_dark')

# Calcula a porcentagem de cada barra
y = df_Setor1['Quantidade']
y_total = y.sum()
y_percent = y / y_total * 100
fig.update_traces(
    text=np.round(y_percent,1).astype(str) + '%', # formata a porcentagem com duas casas decimais e o símbolo "%"
    texttemplate='%{text:.s}', # formata o texto com duas casas decimais
    textposition='outside', # posiciona o texto dentro da barra
)
fig.show(width=1000, height=1000)

In [ ]:
df_Tamanho = df.groupby('Tamanho da empresa').size().sort_values(ascending=False).to_frame()
total = df_Tamanho.sum()[0]
df_Tamanho['percentage'] = df_Tamanho.apply(lambda x: (x/total) * 100, axis=1)
df_Tamanho['percentage'] = df_Tamanho['percentage'].apply(lambda x: '{:.2f}%'.format(x))
df_Tamanho = df_Tamanho.reset_index()
df_Tamanho.columns = ["Tamanho da empresa","Quantidade", "percentage"]
df_Tamanho

Tamanho da empresa  Quantidade percentage
0             +10001          28     38.36%
1        1000 a 5000          17     23.29%
2          200 a 500           8     10.96%
3       5000 a 10000           8     10.96%
4         500 a 1000           4      5.48%
5           51 a 200           4      5.48%
6    Sem informações           3      4.11%
7             2 a 10           1      1.37%

conseguimos perceber que 38.36% das empresas com 10.000 mil ou mais funcionarios.

também 23.29% - 1.000 a 5.000.  

Dar para perceber tem uma tendência de grande empresas utilizarem a ferramentas

In [ ]:
# Calcula a porcentagem de cada barra
y = df_Tamanho['Quantidade']
y_total = y.sum()
y_percent = y / y_total * 100

plt.style.use('dark_background')
fig = px.histogram(df_Tamanho, x='Tamanho da empresa' , y= 'Quantidade')
fig.update_layout(template='plotly_dark')

fig.update_traces(
    text=np.round(y_percent,2).astype(str) + '%', # formata a porcentagem com duas casas decimais e o símbolo "%"
    texttemplate='%{text:.s}', # formata o texto com duas casas decimais
    textposition='outside' # posiciona o texto dentro da barra
)
fig.show(width=1000, height=900)



*   Ponto de atenção, 4,11% dos clientes estão com os dados faltantes.




In [ ]:
df_TamanhoP = df.groupby(['Tamanho da empresa','Produto', 'Tipo Plano']).size().sort_values(ascending=False).to_frame()
total = df_TamanhoP.sum()[0]

df_TamanhoP['percentage'] = df_TamanhoP.apply(lambda x: (x/total) * 100, axis=1)
df_TamanhoP['percentage'] = df_TamanhoP['percentage'].apply(lambda x: '{:.2f}%'.format(x))
df_TamanhoP = df_TamanhoP.reset_index()
df_TamanhoP.columns = ["Tamanho da empresa", 'Produto', 'Tipo Plano', "Quantidade", "percentage"]

fig = px.scatter(df_TamanhoP, x="Tamanho da empresa", y="Produto", size="Quantidade", color="Tipo Plano", size_max=60)
fig.update_layout(template='plotly_dark')
fig.show()

HR Bot não apresentar um padrão, porém Spark nas empresas menos 2 a 1.000 funcionários tendem a optar pelo plano talent, e a parti 5.000 funcionários plano colab começa subir junto com talent, 

In [ ]:
df_Tempo_Vida = df.groupby('Idade da Empresa (anos)').size().sort_values(ascending=False).to_frame()
total = df_Tempo_Vida.sum()[0]
df_Tempo_Vida['percentage'] = df_Tempo_Vida.apply(lambda x: (x/total) * 100, axis=1)
df_Tempo_Vida['percentage'] = df_Tempo_Vida['percentage'].apply(lambda x: '{:.2f}%'.format(x))
df_Tempo_Vida = df_Tempo_Vida.reset_index()
df_Tempo_Vida.columns = ["Idade da Empresa (anos)","Quantidade", "percentage"]

# Create a new column with the decade of each age
df_Tempo_Vida['Decade'] = pd.cut(df_Tempo_Vida['Idade da Empresa (anos)'], bins=[0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 117],
                               labels=[ '1 a 9', '10 a 19', '20 a 29', '30 a 39', '40 a 49', '50 a 59', '60 a 69', '70 a 79', '80 a 89', '90 a 99' ,'+100'])

df_decades = df_Tempo_Vida.groupby('Decade')['Quantidade'].sum().to_frame()

df_decades.sort_values

total = df_decades.sum()[0]
df_decades['percentage'] = df_decades.apply(lambda x: (x/total) * 100, axis=1)
df_decades['percentage'] = df_decades['percentage'].apply(lambda x: '{:.2f}%'.format(x))
df_decades = df_decades.reset_index()
df_decades

Decade  Quantidade percentage
0     1 a 9          13     18.84%
1   10 a 19          21     30.43%
2   20 a 29           8     11.59%
3   30 a 39           7     10.14%
4   40 a 49           5      7.25%
5   50 a 59          14     20.29%
6   60 a 69           0      0.00%
7   70 a 79           0      0.00%
8   80 a 89           0      0.00%
9   90 a 99           0      0.00%
10     +100           1      1.45%

In [ ]:
# Create a bar chart showing the total number of companies in each decade
fig = px.bar(df_decades, x='Decade', y='Quantidade', title='Total number of companies by decade')
fig.update_layout(template='plotly_dark')

# Calcula a porcentagem de cada barra
y = df_decades['Quantidade']
y_total = y.sum()
y_percent = y / y_total * 100

fig.update_traces(
    text=np.round(y_percent,2).astype(str) + '%', # formata a porcentagem com duas casas decimais e o símbolo "%"
    texttemplate='%{text:.s}', # formata o texto com duas casas decimais
    textposition='outside' # posiciona o texto dentro da barra
)

fig.show(width=1000, height=900)


como podemos ver temos uma distribuição de empresas de 1 a 59 anos

In [ ]:
df_total = df.groupby(['Produto', 'Tipo Plano', 'Estado']).size().sort_values(ascending=False).to_frame()


total = df_Setor.sum()[0]
df_total['percentage'] = df_total.apply(lambda x: (x/total) * 100, axis=1)
df_total['percentage'] = df_total['percentage'].apply(lambda x: '{:.2f}%'.format(x))
df_total

0 percentage
Produto Tipo Plano Estado               
Spark   Talent     SP      30     41.10%
        Colab      SP       9     12.33%
HR Bot  Talent     SP       5      6.85%
Spark   Talent     RJ       3      4.11%
                   MG       3      4.11%
                   DF       3      4.11%
        Colab      RS       2      2.74%
HR Bot  Colab      SP       2      2.74%
                   nan      1      1.37%
Spark   Talent     GO       1      1.37%
                   SC       1      1.37%
                   RS       1      1.37%
                   PR       1      1.37%
                   PE       1      1.37%
                   MT       1      1.37%
                   CA       1      1.37%
HR Bot  Talent     MG       1      1.37%
Spark   Colab      SC       1      1.37%
                   PA       1      1.37%
HR Bot  Talent     SC       1      1.37%
                   RJ       1      1.37%
                   PR       1      1.37%
                   MT       1      1.37%
        Colab      GO       1      1.37%

#Dados dos clientes que saíram da base

In [ ]:
# pegando todas as empresas com data de saída
não_zero = df['Data de Saída'] != 0
df_Saida = df.loc[não_zero]
não_zero = (df['Data de Saída'] != 0) #& (df['Data de Saída'] != x)
df_Saida = df.loc[não_zero]

df_Saida

Produto Data de Entrada        Data de Saída    Tipo Saída Tipo Plano  \
7    Spark      2021-05-21  2022-05-22 00:00:00    Voluntário     Talent   
12   Spark      2021-03-21  2022-01-22 00:00:00    Voluntário     Talent   
21  HR Bot      2021-06-21  2022-04-22 00:00:00    Voluntário      Colab   
32   Spark      2021-06-21  2022-10-22 00:00:00    Voluntário     Talent   
33   Spark      2021-10-21  2022-07-22 00:00:00    Voluntário      Colab   
42   Spark      2021-12-21  2022-08-22 00:00:00  Involuntário     Talent   
43   Spark      2021-10-21  2022-10-22 00:00:00    Voluntário     Talent   
53   Spark      2021-09-21  2022-05-22 00:00:00  Involuntário      Colab   
54   Spark      2021-03-21  2022-09-22 00:00:00  Involuntário     Talent   
65   Spark      2021-02-21  2022-03-22 00:00:00  Involuntário     Talent   
66   Spark      2021-05-21  2022-01-22 00:00:00    Voluntário      Colab   

   Natureza Jurídica        Data Abertura  \
7              EDExt  2013-11-21 00:00:00   
12               SEL  1966-07-21 00:00:00   
21               SAF  2015-03-26 00:00:00   
32               SAA  2011-01-31 00:00:00   
33               SAA  1985-08-22 00:00:00   
42               SAA  1973-09-28 00:00:00   
43               SAA  1997-12-23 00:00:00   
53               SEL  2010-04-08 00:00:00   
54               SEL  2010-02-08 00:00:00   
65               SAF  1987-10-09 00:00:00   
66               SAA  2007-11-14 00:00:00   

                        Atividade Econômica Principal  \
7   DESENVOLVIMENTO DE PROGRAMAS DE COMPUTADOR SOB...   
12  FABRICAÇÃO DE PEÇAS E ACESSÓRIOS PARA O SISTEM...   
21  COMÉRCIO ATACADISTA DE COSMÉTICOS E PRODUTOS D...   
32           HOLDINGS DE INSTITUIÇÕES NÃO-FINANCEIRAS   
33  FABRICAÇÃO DE ARTIGOS DO VESTUÁRIO, PRODUZIDOS...   
42                 LOCAÇÃO DE AUTOMÓVEIS SEM CONDUTOR   
43  TRATAMENTO DE DADOS, PROVEDORES DE SERVIÇOS DE...   
53  PORTAIS, PROVEDORES DE CONTEÚDO E OUTROS SERVI...   
54  DESENVOLVIMENTO DE PROGRAMAS DE COMPUTADOR SOB...   
65  DESENVOLVIMENTO DE PROGRAMAS DE COMPUTADOR SOB...   
66                 COMPRA E VENDA DE IMÓVEIS PRÓPRIOS   

                                               Setor Tamanho da empresa  ...  \
7                        Desenvolvimento de software          200 a 500  ...   
12  Fabricação de equipamentos de defesa e espaciais          200 a 500  ...   
21         Fabricação de produtos de higiene pessoal             +10001  ...   
32                        Esportes para espectadores       5000 a 10000  ...   
33                                            Varejo       5000 a 10000  ...   
42                           Planejamento de viagens             +10001  ...   
43                 Tecnologia, Informação e Internet        1000 a 5000  ...   
53                 Tecnologia, Informação e Internet        1000 a 5000  ...   
54                 Tecnologia, Informação e Internet           51 a 200  ...   
65                      Serviços e consultoria de TI             +10001  ...   
66                                 Setor imobiliário        1000 a 5000  ...   

   Idade da Empresa (anos)  Tempo Permanência Ativa (meses)   Status  \
7                        9                               12  Inativo   
12                      56                               10  Inativo   
21                       8                               10  Inativo   
32                      12                               16  Inativo   
33                      37                                9  Inativo   
42                      49                                8  Inativo   
43                      25                               12  Inativo   
53                      13                                8  Inativo   
54                      13                               18  Inativo   
65                      35                               13  Inativo   
66                      15                                8  Inativo   

            Cidade Estado   Latitude  

# Conclusão 

Com base nas informações fornecidas, pode-se criar uma persona com as seguintes características:



*   As empressa tendo a **produto** Spark como o mais popular, com 82% das aquisições;
*   Mais de 75% dos utilizadores possuem o **plano** Talent;
*   A maior parte dos usuários 63,01% está localizada no **estado** de SP;
*  A **cidade** de São Paulo também é a que possui o maior número de usuários, com 42,47%;
*   Os **setores** de tecnologia em geral, é que possui mais usuários 19,18%;
*   A maioria dos utilizadores pertence a empresas com mais de 10.000 **funcionários** 38.36% mas em geral são empresas grandes portes;
*   A maior base dos empresas tem entre 10 a 19	anos 30.43%, porém podemos dizer que a distribuição empresas entre 1 a 59 **anos de idades**;

Essas informações podem ser úteis para direcionar a estratégia de marketing e desenvolvimento de produtos da empresa, além de ajudar a definir o público-alvo em campanha




# Analisando possível correlação.

acredito que melhor opção seja Tempo Permanência Ativa (meses), quanto mais tempo tiver melhor.


*   acredito que melhor opção seja Tempo Permanência Ativa (meses), quanto mais tempo tiver melhor.

*   que as empresas com mais tempo de permanência tem em comum com as outras ?


In [ ]:
df.head(1)

Produto Data de Entrada Data de Saída Tipo Saída Tipo Plano  \
0  HR Bot      2021-03-21             0          0      Colab   

  Natureza Jurídica        Data Abertura  \
0              EIRL  2016-04-20 00:00:00   

                       Atividade Econômica Principal  \
0  SERVIÇOS DE LANTERNAGEM OU FUNILARIA E PINTURA...   

                       Setor Tamanho da empresa  ... Idade da Empresa (anos)  \
0  Administração de serviços           51 a 200  ...                       7   

   Tempo Permanência Ativa (meses)  Status     Cidade Estado   Latitude  \
0                               22   Ativo  GUARULHOS     SP -23.468506   

   Longitude  População  Tamanho da empresa old  Tamanho da empresa cod  
0 -46.531084  1392121.0     51-200 funcionários                     125  

[1 rows x 22 columns]

algumas colunas não será utilizada , olhando datafrane a cima conseguimos determinar as colunas que será excluir.



In [ ]:
df.columns

Index(['Produto', 'Data de Entrada', 'Data de Saída', 'Tipo Saída',
       'Tipo Plano', 'Natureza Jurídica', 'Data Abertura',
       'Atividade Econômica Principal', 'Setor', 'Tamanho da empresa',
       'Unnamed: 17', 'CNPJ cod', 'Idade da Empresa (anos)',
       'Tempo Permanência Ativa (meses)', 'Status', 'Cidade', 'Estado',
       'Latitude', 'Longitude', 'População', 'Tamanho da empresa old',
       'Tamanho da empresa cod'],
      dtype='object')

agora temos outro desafios que é transforma as coluna tipo texto em numérica

In [ ]:
# fazendo uma copia dataframe
df_Numerica = df
# pagando as colunas
df_Numerica = df_Numerica.drop(columns=[
        'Natureza Jurídica', 'Data Abertura',
       'Atividade Econômica Principal', 
       'Unnamed: 17', 'CNPJ cod',
        'Status',
       'Latitude', 'Longitude', 'População', 'Tamanho da empresa old',
       'Tamanho da empresa cod'])
df_Numerica2 = df_Numerica
df_Numerica

Produto Data de Entrada Data de Saída Tipo Saída Tipo Plano  \
0   HR Bot      2021-03-21             0          0      Colab   
1    Spark      2022-10-22             0          0     Talent   
2   HR Bot      2021-01-21             0          0     Talent   
3   HR Bot      2022-08-22             0          0     Talent   
4    Spark      2021-01-21             0          0     Talent   
..     ...             ...           ...        ...        ...   
68   Spark      2022-07-22             0          0     Talent   
69   Spark      2021-03-21             0          0     Talent   
70   Spark      2021-07-21             0          0     Talent   
71   Spark      2022-09-22             0          0     Talent   
72   Spark      2022-11-22             0          0      Colab   

                                                Setor Tamanho da empresa  \
0                           Administração de serviços           51 a 200   
1                                          Manufatura             +10001   
2                                             Seguros          200 a 500   
3                                         Agricultura       5000 a 10000   
4                     Serviços de alimentos e bebidas             +10001   
..                                                ...                ...   
68                                         Manufatura             +10001   
69                                             Varejo             +10001   
70                                   Telecomunicações             +10001   
71  Transporte, Logística, Cadeia de Suprimentos e...       5000 a 10000   
72                                         Manufatura             +10001   

    Idade da Empresa (anos)  Tempo Permanência Ativa (meses)     Cidade Estado  
0                         7                               22  GUARULHOS     SP  
1                       116                                3  SAO PAULO     SP  
2                        13                               24  SAO PAULO     SP  
3                        46                                5     CUIABA     MT  
4                        17                               24  SAO PAULO     SP  
..                      ...                              ...        ...    ...  
68                       54                                6  SAO PAULO     SP  
69                       14                               22  SAO PAULO     SP  
70                       25                               18  SAO PAULO     SP  
71                       12                                4  SAO PAULO     SP  
72                       56                                2  SAO PAULO     SP  

[73 rows x 11 columns]

As colunas "Produto", "Tipo Saída", "Tipo Plano", "Setor", "Cidade" e "Estado" são todas variáveis ​​qualitativas. As variáveis ​​qualitativas podem ser divididas em dois tipos: nominais e ordinais. As variáveis ​​nominais são aquelas que representam categorias sem nenhuma ordenação implícita, enquanto as variáveis ​​ordinais representam categorias com uma ordem implícita.


---



Nesse caso, os "Produto", "Tipo Saída", "Tipo Plano" e "Setor" são variáveis **​​qualitativas nominais**, pois não há ordenação implícita nas categorias que representam. 

Já as colunas "Cidade" e "Estado" são variáveis ​​**qualitativas ordinais**, pois há uma ordem implícita nas categorias que representam (por exemplo, as cidades podem ser ordenadas alfabeticamente ou por região geográfica).

são variáveis ​​qualitativas, pois representam categorias ou atributos que não podem ser medidos numericamente. 





In [ ]:
# Depois disso,  "Produto", "Tipo Saída", "Tipo Plano" e "Setor" serão transformadas em variáveis ​​nominais, onde os valores serão representados por números inteiros.
df_Numerica["Produto"] = df_Numerica["Produto"].factorize()[0]
df_Numerica["Tipo Saída"] = df_Numerica["Tipo Saída"].factorize()[0]
df_Numerica["Tipo Plano"] = df_Numerica["Tipo Plano"].factorize()[0]
df_Numerica["Setor"] = df_Numerica["Setor"].factorize()[0]

Para transformar as colunas "Produto", "Tipo Saída", "Tipo Plano" e "Setor" em variáveis ​​qualitativas nominais para análise, você pode utilizar o método "fatorizar" dos pandas. Esse método converte os valores dessas colunas em números inteiros, atribuindo um número para cada valor único da coluna.

In [ ]:

df_coluna = pd.DataFrame(df_Numerica[["Cidade", "Estado"]])
df_coluna

Cidade Estado
0   GUARULHOS     SP
1   SAO PAULO     SP
2   SAO PAULO     SP
3      CUIABA     MT
4   SAO PAULO     SP
..        ...    ...
68  SAO PAULO     SP
69  SAO PAULO     SP
70  SAO PAULO     SP
71  SAO PAULO     SP
72  SAO PAULO     SP

[73 rows x 2 columns]

Para transformar as colunas "Cidade" e "Estado" em variáveis qualitativas ordinais, pode se utilizar o método "factorize" do pandas, assim como foi utilizado no código anterior. No entanto, é **importante lembrar** que, ao fazer isso, os valores dessas colunas serão convertidos em números inteiros de forma arbitrária, ou seja, não haverá relação entre os valores numéricos atribuídos e a ordem real das cidades ou estados.

---

Se você deseja que os valores numéricos reflitam a ordem das cidades ou estados de alguma forma (por exemplo, se você quiser que as cidades ou estados sejam classificados de acordo com o tamanho da população), será necessário criar uma tabela adicional com os valores das cidades ou estados e atribuir os valores numéricos de acordo com essa tabela.



---


Dessas forma tiver que fazer algumas pequenas modificações pra que eu tenha a coluna com nome cidade e estando junto coluna numérica

In [ ]:
# Pegar coluna cidade e estado fazer uma cópia em outro df vazio
df_coluna = pd.DataFrame(df_Numerica[["Cidade", "Estado"]])
df_coluna
# serão transformadas em variáveis qualitativas ordinais, onde os valores serão representados por números inteiros.
df_coluna["Cidade"] = df_coluna["Cidade"].factorize()[0]
df_coluna["Estado"] = df_coluna["Estado"].factorize()[0]
df_coluna = df_coluna.rename(columns={"Cidade": "Cidade_Numerica", "Estado": "Estado_Numerica"})
df_coluna
#Jjunta dois df
df_Numerica = pd.concat([df_Numerica, df_coluna], axis=1, join='inner')

df_Numerica

Produto Data de Entrada Data de Saída  Tipo Saída  Tipo Plano  Setor  \
0         0      2021-03-21             0           0           0      0   
1         1      2022-10-22             0           0           1      1   
2         0      2021-01-21             0           0           1      2   
3         0      2022-08-22             0           0           1      3   
4         1      2021-01-21             0           0           1      4   
..      ...             ...           ...         ...         ...    ...   
68        1      2022-07-22             0           0           1      1   
69        1      2021-03-21             0           0           1     21   
70        1      2021-07-21             0           0           1     32   
71        1      2022-09-22             0           0           1     33   
72        1      2022-11-22             0           0           0      1   

   Tamanho da empresa  Idade da Empresa (anos)  \
0            51 a 200                        7   
1              +10001                      116   
2           200 a 500                       13   
3        5000 a 10000                       46   
4              +10001                       17   
..                ...                      ...   
68             +10001                       54   
69             +10001                       14   
70             +10001                       25   
71       5000 a 10000                       12   
72             +10001                       56   

    Tempo Permanência Ativa (meses)     Cidade Estado  Cidade_Numerica  \
0                                22  GUARULHOS     SP                0   
1                                 3  SAO PAULO     SP                1   
2                                24  SAO PAULO     SP                1   
3                                 5     CUIABA     MT                2   
4                                24  SAO PAULO     SP                1   
..                              ...        ...    ...              ...   
68                                6  SAO PAULO     SP                1   
69                               22  SAO PAULO     SP                1   
70                               18  SAO PAULO     SP                1   
71                                4  SAO PAULO     SP                1   
72                                2  SAO PAULO     SP                1   

    Estado_Numerica  
0                 0  
1                 0  
2                 0  
3                 1  
4                 0  
..              ...  
68                0  
69                0  
70                0  
71                0  
72                0  

[73 rows x 13 columns]

As colunas "Data de Entrada", "Data de Saída", "Tamanho da empresa", "Idade da Empresa (anos)" e "Tempo Permanência Ativa (meses)" são todas variáveis ​​quantitativas. As variáveis ​​quantitativas podem ser divididas em dois tipos: discretas e contínuas. As variáveis ​​discretas só podem assumir valores discretos, enquanto as variáveis ​​contínuas podem assumir qualquer valor dentro de um intervalo.


---


Nesse caso, os "Tamanho da empresa" e "Tempo Permanência Ativa (meses)" são variáveis ​​**quantitativas discretas,** pois só podem assumir valores discretos. 

Já as colunas "Data de Entrada" e "Data de Saída" são **quantitativas contínuas**, pois podem assumir qualquer valor dentro de um intervalo (por exemplo, a data de entrada pode ser "2021-03-21", mas também pode ser "2021-03-21 12:34:56"). 

A coluna "Idade da Empresa (anos)" também é uma variável quantitativa discreta, pois só pode assumir valores inteiros (por exemplo, a idade da empresa pode ser "7", mas não pode ser "7,5").





In [ ]:
df_Numerica.dtypes

Produto                                     int64
Data de Entrada                    datetime64[ns]
Data de Saída                              object
Tipo Saída                                  int64
Tipo Plano                                  int64
Setor                                       int64
Tamanho da empresa                         object
Idade da Empresa (anos)                     Int64
Tempo Permanência Ativa (meses)             Int64
Cidade                                     object
Estado                                     object
Cidade_Numerica                             int64
Estado_Numerica                             int64
dtype: object

In [ ]:
df_Numerica.columns

Index(['Produto', 'Data de Entrada', 'Data de Saída', 'Tipo Saída',
       'Tipo Plano', 'Setor', 'Tamanho da empresa', 'Idade da Empresa (anos)',
       'Tempo Permanência Ativa (meses)', 'Cidade', 'Estado',
       'Cidade_Numerica', 'Estado_Numerica'],
      dtype='object')

In [ ]:
#df_Numerica =  pd.get_dummies(df_Numerica, columns= ['Produto' ,'Tipo Plano', 'Setor',
    #   'Tamanho da empresa', 'Idade da Empresa (anos)',
   #     'Cidade', 'Estado',])
#df_Numerica

A coluna pode ser transformada usando One-Hot-Encoding

In [ ]:
corelacao = df_Numerica._get_numeric_data()
# Calculate the correlations between the columns
new_correlations = corelacao.corr()
# Select the columns with a correlation greater than 0.55 or less than -0.55
highly_correlated_features = new_correlations[(new_correlations > 0.55) | (new_correlations < -0.55)]
# Substituir os valores inferior por '-'
highly_correlated_features.fillna('-')

Produto Tipo Saída Tipo Plano Setor  \
Produto                             1.0          -          -     -   
Tipo Saída                            -        1.0          -     -   
Tipo Plano                            -          -        1.0     -   
Setor                                 -          -          -   1.0   
Idade da Empresa (anos)               -          -          -     -   
Tempo Permanência Ativa (meses)       -          -          -     -   
Cidade_Numerica                       -          -          -     -   
Estado_Numerica                       -          -          -     -   

                                Idade da Empresa (anos)  \
Produto                                               -   
Tipo Saída                                            -   
Tipo Plano                                            -   
Setor                                                 -   
Idade da Empresa (anos)                             1.0   
Tempo Permanência Ativa (meses)                       -   
Cidade_Numerica                                       -   
Estado_Numerica                                       -   

                                Tempo Permanência Ativa (meses)  \
Produto                                                       -   
Tipo Saída                                                    -   
Tipo Plano                                                    -   
Setor                                                         -   
Idade da Empresa (anos)                                       -   
Tempo Permanência Ativa (meses)                             1.0   
Cidade_Numerica                                               -   
Estado_Numerica                                               -   

                                Cidade_Numerica Estado_Numerica  
Produto                                       -               -  
Tipo Saída                                    -               -  
Tipo Plano                                    -               -  
Setor                                         -               -  
Idade da Empresa (anos)                       -               -  
Tempo Permanência Ativa (meses)               -               -  
Cidade_Numerica                             1.0               -  
Estado_Numerica                               -             1.0

In [ ]:

fig = px.imshow(new_correlations,
                text_auto=".2f",
                
                x=new_correlations.columns,
                y=new_correlations.index,
                color_continuous_scale='Viridis',
                title='Correlação entre as variáveis')

fig.update_layout(width=1000, height=1000)
fig.update_layout(template='plotly_dark')
fig.show()

In [ ]:
# Pegando todas a colunas e criando uma lista para armazenar 
columns = df_Numerica.columns.tolist()
# laços vai percorrer toda a 
lista com nome de cada coluna
for var in columns:
    fig = px.scatter(df_Numerica, x=var, y='Tempo Permanência Ativa (meses)')
    fig.update_layout(template='plotly_dark', width=1500, height=400)
    
    fig.show()

In [ ]:
df_Numerica2 =  pd.get_dummies(df_Numerica, columns= ['Produto' ,'Tipo Plano', 'Setor',
     'Tamanho da empresa', 'Idade da Empresa (anos)',
     'Cidade', 'Estado',])
df_Numerica2

corelacao2 = df_Numerica2._get_numeric_data()
# Calculate the correlations between the columns
new_correlations2 = corelacao2.corr()

corr_filtered = new_correlations2.where((new_correlations2 > 0.55) | (new_correlations2 < -0.55))


fig = px.imshow(corr_filtered,
                text_auto=".2f",
                
                x=corr_filtered.columns,
                y=corr_filtered.index,
                color_continuous_scale='Viridis',
                title='Correlação entre as variáveis')

fig.update_layout(width=1000, height=1000)
fig.update_layout(template='plotly_dark')
fig.show()

não encontrei correlação entre Tempo Permanência Ativa (meses) 

# Carregar 2 bases de dados Nps - Net Promoter Score.

In [ ]:
df_Nps = pd.read_csv('/content/drive/MyDrive/Projeto 17/NPS - CS.csv')
df_Nps_produto = pd.read_csv("/content/drive/MyDrive/Projeto 17/NPS - Produto.csv")


In [ ]:
df_Nps.head(5)

="user"  \
0  ="d92ec849-e1a0-6d5f-9211-7cd017923b45"   
1  ="901e26b6-7892-63cb-bb79-4ff4f74ce5b0"   
2  ="35cd4df4-f77e-7776-2295-c4a7043dd0d8"   
3  ="35cd4df4-f77e-7776-2295-c4a7043dd0d8"   
4  ="35cd4df4-f77e-7776-2295-c4a7043dd0d8"   

                                ="channel"          ="LGPD"  \
0  ="d92ec849-e1a0-6d5f-9211-7cd017923b45"  ="Sim, aceito!"   
1  ="901e26b6-7892-63cb-bb79-4ff4f74ce5b0"  ="Sim, aceito!"   
2  ="35cd4df4-f77e-7776-2295-c4a7043dd0d8"   ="Não aceito!"   
3  ="35cd4df4-f77e-7776-2295-c4a7043dd0d8"  ="Sim, aceito!"   
4  ="35cd4df4-f77e-7776-2295-c4a7043dd0d8"  ="Sim, aceito!"   

                     ="nome_"                              ="Email" ="email_"  \
0  ="Vinícius Rodrigues Leal"   ="vinicius.leal@matchboxbrasil.com"     ="10"   
1  ="Vinicius Rodrigues Leal"              ="vinicius@vinicius.com"     ="10"   
2                         NaN                                   NaN       NaN   
3                ="gabrielle"  ="gabrielle.tuao@matchboxbrasil.com"     ="10"   
4                ="Gabrielle"        ="gabrielle.tuao@hrbot.com.br"     ="10"   

                                    ="Justificativa"    ="_status"  \
0  ="Estou muito feliz com o atendimento da área ...  ="completed"   
1                                           ="teste"  ="completed"   
2                                                NaN  ="completed"   
3                                           ="Teste"  ="completed"   
4                                           ="Teste"  ="completed"   

  ="Nome_da_Empresa" ="Nota_NP" ="Nome_da_Empresa".1 ="Nota_NPS"  \
0                NaN        NaN                  NaN         NaN   
1                NaN        NaN                  NaN         NaN   
2                NaN        NaN                  NaN         NaN   
3                NaN        NaN                  NaN         NaN   
4                NaN        NaN                  NaN         NaN   

  ="Nome_da_Empres" ="Nota_NPS".1 ="Data de aceitação LGPD"  
0               NaN           NaN       ="14/04/2022 18:18"  
1               NaN           NaN       ="14/04/2022 18:24"  
2               NaN           NaN       ="25/04/2022 16:12"  
3               NaN           NaN       ="25/04/2022 17:42"  
4               NaN           NaN       ="25/04/2022 17:46"

Essa base esta muito poluída e com muitos dados nulos.

pelo que percebi foram feitos 3 nps com intervalos 6 mês entre eles, então 1 ano e meio esse foi resultados.

Infelizmente a empresa precisa utilizar outros planejamentos para conseguir melhores resultados no **nps**

In [ ]:
df_Nps.isna().mean()

="user"                      0.000000
="channel"                   0.000000
="LGPD"                      0.878049
="nome_"                     0.780488
="Email"                     0.902439
="email_"                    0.780488
="Justificativa"             0.219512
="_status"                   0.000000
="Nome_da_Empresa"           0.756098
="Nota_NP"                   0.756098
="Nome_da_Empresa".1         0.975610
="Nota_NPS"                  0.975610
="Nome_da_Empres"            0.707317
="Nota_NPS".1                0.707317
="Data de aceitação LGPD"    0.000000
dtype: float64

In [ ]:
df_Nps = df_Nps.apply(lambda x: x.str.lower() if x.dtype == "object" else x)

df_Nps.describe(include='all', datetime_is_numeric=True).T

count unique  \
="user"                      41     35   
="channel"                   41     35   
="LGPD"                       5      2   
="nome_"                      9      6   
="Email"                      4      4   
="email_"                     9      2   
="Justificativa"             32     23   
="_status"                   41      1   
="Nome_da_Empresa"           10      5   
="Nota_NP"                   10      2   
="Nome_da_Empresa".1          1      1   
="Nota_NPS"                   1      1   
="Nome_da_Empres"            12      7   
="Nota_NPS".1                12      3   
="Data de aceitação LGPD"    41     38   

                                                               top freq  
="user"                    ="8fa5d658-0233-d522-ea8f-51453775fa98"    3  
="channel"                 ="8fa5d658-0233-d522-ea8f-51453775fa98"    3  
="LGPD"                                            ="sim, aceito!"    4  
="nome_"                                                  ="teste"    3  
="Email"                       ="vinicius.leal@matchboxbrasil.com"    1  
="email_"                                                    ="10"    8  
="Justificativa"                                          ="teste"   10  
="_status"                                            ="completed"   41  
="Nome_da_Empresa"                                     ="matchbox"    5  
="Nota_NP"                                                   ="10"    7  
="Nome_da_Empresa".1                                      ="teste"    1  
="Nota_NPS"                                                  ="10"    1  
="Nome_da_Empres"                                        ="idwall"    3  
="Nota_NPS".1                                                 ="9"    5  
="Data de aceitação LGPD"                      ="15/06/2022 21:00"    4

In [ ]:
df_Nps_produto

="user"  \
0   ="7cc626a8-a81f-9ff7-8f09-4684cbb12db4"   
1   ="1bf8717f-4b38-8df2-7660-dd754391d980"   
2   ="17ffbacd-f07b-4a32-a173-1a55ae2fd69b"   
3   ="b0bbae3b-f9e5-867c-3f2e-bb48164c69d2"   
4   ="1dd9bbc6-24f1-f7bd-906d-37740a589e77"   
5   ="689bde89-0dfa-7ee8-70be-28ca8ac90970"   
6   ="65ca010e-276d-5c09-d761-31865094e774"   
7   ="0758a17a-439b-fae7-9ecc-41f417b0b71b"   
8   ="e72d42c5-9e02-e788-2ee5-9ead6832439c"   
9   ="d87268d4-bacb-8dc0-4ca1-8eac7ac6554a"   
10  ="9258ec59-9b33-1437-4a30-d21c0471654f"   

                                 ="channel"  ="Nome_da_Empre" ="Nota_NPS"  \
0   ="7cc626a8-a81f-9ff7-8f09-4684cbb12db4"          ="teste"       ="10"   
1   ="1bf8717f-4b38-8df2-7660-dd754391d980"         ="teste~"        ="4"   
2   ="17ffbacd-f07b-4a32-a173-1a55ae2fd69b"         ="idwall"        ="6"   
3   ="b0bbae3b-f9e5-867c-3f2e-bb48164c69d2"          ="Ambev"       ="10"   
4   ="1dd9bbc6-24f1-f7bd-906d-37740a589e77"         ="Ambev "        ="9"   
5   ="689bde89-0dfa-7ee8-70be-28ca8ac90970"      ="mat5chbox"        ="9"   
6   ="65ca010e-276d-5c09-d761-31865094e774"  ="Mercado Livre"       ="10"   
7   ="0758a17a-439b-fae7-9ecc-41f417b0b71b"         ="Raízen"        ="9"   
8   ="e72d42c5-9e02-e788-2ee5-9ead6832439c"         ="idwall"        ="9"   
9   ="d87268d4-bacb-8dc0-4ca1-8eac7ac6554a"         ="idwall"        ="7"   
10  ="9258ec59-9b33-1437-4a30-d21c0471654f"         ="teste~"       ="10"   

                                     ="Justificativa"    ="_status"  \
0                                            ="teste"  ="completed"   
1                                              ="eee"  ="completed"   
2   ="Temos enfrentado diversas restrições e limit...  ="completed"   
3   ="Ferramenta ajuda na otimização e automatizaç...  ="completed"   
4   ="Ótima ferramenta! Facilita a rotina através ...  ="completed"   
5                                            ="teste"  ="completed"   
6   ="Uma inovação para o RH. De forma simples con...  ="completed"   
7   ="Plataforma que nos ajuda no processo de atra...  ="completed"   
8   ="ferramenta muito boa, possui apenas alguns i...  ="completed"   
9   ="Existem oportunidades de melhorias e precisa...  ="completed"   
10                                           ="teste"  ="completed"   

   ="Data de finalização da Conversa"  
0                 ="22/09/2022 16:58"  
1                 ="06/10/2022 14:16"  
2                 ="06/10/2022 14:48"  
3                 ="06/10/2022 15:05"  
4                 ="07/10/2022 14:31"  
5                 ="12/12/2022 14:38"  
6                 ="12/12/2022 16:09"  
7                 ="12/12/2022 16:35"  
8                 ="19/12/2022 16:37"  
9                 ="20/12/2022 17:25"  
10                ="21/12/2022 11:01"

In [ ]:
df_Nps_produto = df_Nps_produto.apply(lambda x: x.str.lower() if x.dtype == "object" else x)
df_Nps_produto.describe(include='all', datetime_is_numeric=True).T

count unique  \
="user"                               11     11   
="channel"                            11     11   
="Nome_da_Empre"                      11      8   
="Nota_NPS"                           11      5   
="Justificativa"                      11      9   
="_status"                            11      1   
="Data de finalização da Conversa"    11     11   

                                                                        top  \
="user"                             ="7cc626a8-a81f-9ff7-8f09-4684cbb12db4"   
="channel"                          ="7cc626a8-a81f-9ff7-8f09-4684cbb12db4"   
="Nome_da_Empre"                                                  ="idwall"   
="Nota_NPS"                                                           ="10"   
="Justificativa"                                                   ="teste"   
="_status"                                                     ="completed"   
="Data de finalização da Conversa"                      ="22/09/2022 16:58"   

                                   freq  
="user"                               1  
="channel"                            1  
="Nome_da_Empre"                      3  
="Nota_NPS"                           4  
="Justificativa"                      3  
="_status"                           11  
="Data de finalização da Conversa"    1

# Modelo Ml


In [ ]:
df_Numerica2.head(3)

Data de Entrada Data de Saída  Tipo Saída  Tempo Permanência Ativa (meses)  \
0      2021-03-21             0           0                               22   
1      2022-10-22             0           0                                3   
2      2021-01-21             0           0                               24   

   Cidade_Numerica  Estado_Numerica  Produto_0  Produto_1  Tipo Plano_0  \
0                0                0          1          0             1   
1                1                0          0          1             0   
2                1                0          1          0             0   

   Tipo Plano_1  ...  Estado_MG  Estado_MT  Estado_PA  Estado_PE  Estado_PR  \
0             0  ...          0          0          0          0          0   
1             1  ...          0          0          0          0          0   
2             1  ...          0          0          0          0          0   

   Estado_RJ  Estado_RS  Estado_SC  Estado_SP  Estado_nan  
0          0          0          0          1           0  
1          0          0          0          1           0  
2          0          0          0          1           0  

[3 rows x 132 columns]


Temos 3 desafios com base informações fornecidos pela empresa.

1.   Mapeamento de Personas; OK
2.   Redução churn;
3.   Aumentar base de clientes pago;

**precisamos saber se é possível criar esses modelos com os dados que empresa fornece.**


**Para criar um modelo de churn**

pode precisar de diversos tipos de dados, dependendo da natureza da sua empresa e do seu negócio. Alguns exemplos de dados que podem ser úteis para prever o churn incluem:

* Dados demográficos do cliente, como idade, gênero, renda e localização geográfica
* Dados sobre o uso do produto ou serviço pelo cliente, como frequência de uso, tempo de uso e valor gasto
* Dados sobre o relacionamento do cliente com a empresa, como atendimentos ao cliente, reclamações e feedback
* Dados sobre a concorrência e o mercado, como preços de produtos similares e ofertas de outras empresas
* Dados sobre eventos externos que podem afetar a decisão do cliente de continuar usando o produto ou serviço, como mudanças nas necessidades do  cliente ou mudanças nas condições do mercado.


---


na matchboxbrasil. 

tenho alguns clientes já saíram da base;

mas como foi comportamento deles? 

deixaram pagamentos atrasar?

utilizaram menos plataforma com tempo?

qual é media de tempo usuário na plataforma e ferramentas?

deram feedback?

como meu modelo vai mensurar?

**Para criar um modelo captar clientes do free e jogar no pago**


* Para aplicar a clusterização.

precisará de um conjunto de dados que inclua informações sobre os elementos que deseja agrupar empresas e as características que deseja usar para agrupá-los (por exemplo, persona que já definimos).


* modelo de recomendação.

Além disso, é importante ter em mente que os modelos de recomendação geralmente dependem de dados de interação dos usuários, como visualizações de produtos ou classificações, para gerar recomendações precisas. Portanto, é importante ter acesso a esses dados de interação para treinar o modelo de recomendação.

Além disso, é útil ter dados de contexto adicionais que podem ajudar a personalizar as recomendações, como informações demográficas dos usuários ou dados de localização. Isso pode ajudar a fornecer recomendações mais relevantes para cada indivíduo.


---
* ação após modelos

Dessa maneira podemos fazer recomendações para clientes que estão usando um plano gratuito é oferecer a eles uma avaliação gratuita do plano pago. Isso lhes dará a oportunidade de experimentar as funcionalidades adicionais que o plano pago oferece e decidir se vale a pena fazer a atualização. Você também pode oferecer descontos ou promoções exclusivas para incentivar os clientes a fazer a atualização.


Outra opção é enviar aos clientes relatórios ou estatísticas mostrando como eles estão se beneficiando do uso do seu produto ou serviço e como o plano pago pode oferecer ainda mais benefícios. Você também pode entrar em contato com os clientes convidados e perguntar-lhes como eles estão usando o produto e se eles gostariam de ver quaisquer novas funcionalidades ou recursos adicionais. Isso pode ajudá-lo a identificar o que os clientes estão procurando e oferecer recomendações personalizadas sobre como o plano pago pode atender às suas necessidades.


collaborative filtering - recomendação de plano

In [ ]:
!pip install --pre pycaret



PyCaret deployment capabilities
PyCaret is a deployment ready library in Python which means all the steps performed in an ML experiment can be reproduced using a pipeline that is reproducible and guaranteed for production.  A pipeline can be saved in a binary file format that is transferable across environments.